### PR TITLE
Implement Justification and finalization Processing

### PR DESCRIPTION
**Pull request by terencechain**
This is part of #2307, it implements justification and finalization processing function from v0.6 spec

Changes:
- Implemented justification and finalization
- Added tests for the above

### CODE REVIEW

**Comment by nisdas** [comment]
@terencechain we already have this here https://github.com/prysmaticlabs/prysm/blob/9a626706eef07c7ffefcb21eab7bb85ca6faa0c1/beacon-chain/core/epoch/epoch_processing.go#L136. Is the function you are writing a replacement ? because they seem pretty similar

**Comment by terencechain** [comment]
> @terencechain we already have this here
> 
> https://github.com/prysmaticlabs/prysm/blob/9a626706eef07c7ffefcb21eab7bb85ca6faa0c1/beacon-chain/core/epoch/epoch_processing.go#L136
> 
> . Is the function you are writing a replacement ? because they seem pretty similar

It's a complete replacement, I'm going to delete the old one when this is all done

File: beacon-chain/core/epoch/epoch_processing.go [hunk 1]
```diff
@@ -386,6 +386,119 @@ func UpdateLatestActiveIndexRoots(state *pb.BeaconState) (*pb.BeaconState, error
...
+//    if (bitfield >> 0) % 8 == 0b111 and old_current_justified_epoch == current_epoch - 2:
+//        state.finalized_epoch = old_current_justified_epoch
+//        state.finalized_root = get_block_root(state, state.finalized_epoch)
+//    # The 1st/2nd most recent epochs are justified, the 1st using the 2nd as source
+//    if (bitfield >> 0) % 4 == 0b11 and old_current_justified_epoch == current_epoch - 1:
+//        state.finalized_epoch = old_current_justified_epoch
+//        state.finalized_root = get_block_root(state, state.finalized_epoch)
+func ProcessJustificationFinalization(state *pb.BeaconState, prevAttestedBal uint64, currAttestedBal uint64) (
+	*pb.BeaconState, error) {
+	// There's no reason to process justification until the 2nd epoch.
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/core/epoch/epoch_processing.go]: Yeah 3rd epoch. Good catch!

**Comment by terencechain** [comment]
> Will you be deleting the old function later ?

Correct. Once we write the new `ProcessEpoch` function, the old `ProcessJustificationFinalization` will be deleted. Replacing/deleting that now will make this hard to review

**Pull request merged**